src/Files.App/Helpers/StorageItemCollection/ConcurrentCollection.cs

```diff
@@ -0,0 +1,294 @@
+
+				collection.RemoveAt(index);
+			}
+
+			return true;
+		}
+
+		public IEnumerator<T> GetEnumerator()
+		{
+			return new BlockingListEnumerator<T>(collection, syncRoot);
+		}
+
+		IEnumerator IEnumerable.GetEnumerator()
+		{
+			return GetEnumerator();
+		}
+
+		public int IndexOf(T? item)
+		{
+			if (item is null)
+				return -1;
+
+			lock (syncRoot)
+			{
+				return collection.IndexOf(item);
+			}
+		}
+
+		public void Insert(int index, T? item)
+		{
+			if (item is null)
+				return;
+
+			lock (syncRoot)
+			{
+				collection.Insert(index, item);
+			}
+		}
+
+		public void RemoveAt(int index)
+		{
+			T item;
+
+			lock (syncRoot)
+			{
+				item = collection[index];
+				collection.RemoveAt(index);
+			}
+		}
+
+		public void AddRange(IEnumerable<T> items)
+		{
+			if (!items.Any())
+				return;
+
+			lock (syncRoot)
+			{
+				collection.AddRange(items);
+			}
+		}
+
+		public void InsertRange(int index, IEnumerable<T> items)
+		{
+			if (!items.Any())
+				return;
+
+			lock (syncRoot)
+			{
+				collection.InsertRange(index, items);
+			}
+		}
+
+		public void RemoveRange(int index, int count)
+		{
+			if (count <= 0)
+				return;
+
+			List<T> items;
+
+			lock (syncRoot)
+			{
+				items = collection.Skip(index).Take(count).ToList();
+				collection.RemoveRange(index, count);
+			}
+		}
+
+		public void ReplaceRange(int index, IEnumerable<T> items)
+		{
+			var count = items.Count();
+
+			if (count == 0)
+				return;
+
+			List<T> oldItems;
+			List<T> newItems;
+
+			lock (syncRoot)
+			{
+				oldItems = collection.Skip(index).Take(count).ToList();
+				newItems = items.ToList();
+				collection.InsertRange(index, newItems);
+				collection.RemoveRange(index + count, count);
+			}
+		}
+
+		public void Sort()
+		{
+			lock (syncRoot)
+			{
+				collection.Sort();
+			}
+		}
+
+		public void Sort(Comparison<T> comparison)
+		{
+			lock (syncRoot)
+			{
+				collection.Sort(comparison);
+			}
+		}
+
+		public void Order(Func<List<T>, IEnumerable<T>> func)
+		{
+			IEnumerable<T> result;
+			lock (syncRoot)
+			{
+				result = func.Invoke(collection);
+			}
+
+			ReplaceRange(0, result);
+		}
+
+		public void OrderOne(Func<List<T>, IEnumerable<T>> func, T item)
+		{
+			IList<T> result;
+			lock (syncRoot)
+			{
+				result = func.Invoke(collection).ToList();
+			}
+
+			Remove(item);
+			var index = result.IndexOf(item);
+			if (index != -1)
+				Insert(index, item);
+		}
+
+		int IList.Add(object? value)
+		{
+			if (value is null)
+				return -1;
+
+			int index;
+
+			lock (syncRoot)
+			{
+				index = ((IList)collection).Add((T)value);
+			}
+
+			return index;
+		}
+
+		bool IList.Contains(object? value) => Contains((T?)value);
+
+		int IList.IndexOf(object? value) => IndexOf((T?)value);
+
+		void IList.Insert(int index, object? value) => Insert(index, (T?)value);
+
+		void IList.Remove(object? value) => Remove((T?)value);
+
+		void ICollection.CopyTo(Array array, int index) => CopyTo((T[])array, index);
```

Comment:
I think you shouldn't use `arrow functions` here
What do you think @0x5bfa

Reply:
We are obliged to avoid its use.